sumologic_collectd_metrics/metrics_converter.py

```diff
@@ -144,3 +155,82 @@ def convert_to_metrics(data, data_set, sep):
         metrics.append(metric)
 
     return metrics
+
+
+def process_signalfx_statsd_tags(data):
+    """
+    Process statsd tags embedded in the metric name within the supplied raw data. This function
+    modifies the raw data in-place.
+
+    :param data:
+    :type data: collectd.Values
+    :return:
+    :rtype: None
+    """
+    cleaned_type_instance, new_metadata = parse_statsd_signalfx_metric_name(
+        data.type_instance
+    )
+    data.type_instance = cleaned_type_instance
+    data.meta.update(new_metadata)
+
+
+def parse_statsd_signalfx_metric_name(metric_name):
+    """
+    Parse the metric name and extract metadata from it according to SignalFx' custom extension to
+    the StatsD protocol.
+    See: https://docs.signalfx.com/en/latest/integrations/agent/monitors/collectd-statsd.html#adding-dimensions-to-statsd-metrics
+
+    :param metric_name: The metric name
+    :type metric_name: str
+    :return: A pair consisting of the modified metric name without the metadata, and a dict of
+    extracted metadata.
+    :rtype: (str, dict)
+    """  # pylint: disable=line-too-long
+    matcher = re.compile(r"\[([^\]]+)\]")
+    matches = matcher.findall(metric_name)
+    if len(matches) > 1:
+        # if we find more than one metadata segment, something is probably wrong, and we should
+        # ignore it
+        logger.warning(
+            "Found two metadata segments in metric name `%s`. "
+            "This is likely a configuration error." % metric_name
+        )
+        return metric_name, {}
+
+    if len(matches) == 0:
+        # if we find no segments, we have nothing to do
+        return metric_name, {}
+
+    metadata = {}
+    tag_string = matches[0]
+    key_value_strings = tag_string.split(",")
+    for key_value_string in key_value_strings:
+        key_value_string = key_value_string.strip()
+        parts = [p.strip() for p in key_value_string.split("=")]
+        if len(parts) != 2:  # ignore malformed entries, but parse as much as we can
+            logger.warning(
+                "Invalid key=value pair `%s` in metric name `%s`, ignoring it"
+                % (key_value_string, metric_name)
+            )
+            continue
+        key, value = parts
+        if key == "":
+            logger.warning(
+                "Empty keys aren't supported, ignoring key=value pair `%s` in metric name `%s`"
+                % (key_value_string, metric_name)
+            )
+            continue
+        if key in metadata:
+            logger.warning(
+                "Key `%s` is defined multiple times in metric name `%s`, ignoring subsequent values"
```

Comment:
Is this the behavior of the SignalFx plugin that we're trying to mimic here? I mean that the value from the first occurrence of the key is taken and not from the last one.

Reply:
I don't actually know, their documentation doesn't say. 😒  @gizmo-rt do you know?

Reply:
Yeah the code at https://github.com/signalfx/signalfx-agent/tree/main/pkg/monitors/collectd/statsd doesn't really have anything in this regard.